site/app/controllers/course/CourseMaterialsController.php

```diff
@@ -438,30 +438,108 @@ public function ajaxEditCourseMaterialsFiles(bool $flush = true): JsonResponse {
         if (isset($_POST['sort_priority'])) {
             $course_material->setPriority($_POST['sort_priority']);
         }
-        if (isset($_POST['link_url']) && isset($_POST['link_title']) && $course_material->isLink()) {
-            if ($_POST['link_title'] !== $course_material->getUrlTitle()) {
-                $path = $course_material->getPath();
-                $dirs = explode("/", $path);
-                array_pop($dirs);
-                $path = implode("/", $dirs);
-                $file_name = urlencode("link-" . $_POST['link_title']);
+
+
+
+        if (isset($_POST['file_path']) || isset($_POST['title'])) {
+            $path = $course_material->getPath();
+            $upload_path = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "uploads", "course_materials");
+            $requested_path = $_POST['file_path'];
+            $new_path = FileUtils::joinPaths($upload_path, $requested_path);
+
+            $lastSlashPos = strrpos($new_path, "/");
+            if (isset($_POST['title'])) {
+                $file_name = $_POST['title'];
+                $new_path = substr($new_path, 0, $lastSlashPos + 1);
+                $new_path = FileUtils::joinPaths($new_path, $file_name);
+            }
```

Comment:
You could use the php function `dirname` to clean this up a bit and not do as many substrings and position checks